src/corona_hakab_model/manager.py

```diff
@@ -137,23 +136,33 @@ def progress_tests_and_isolation(self, new_tests: List[PendingTestResult]):
         new_results = self.pending_test_results.advance()
         for agent, test_result, _ in new_results:
             if test_result:
-                if not self.ever_tested_positive_vector[agent.index]:
+                if not self.tested_positive_vector_counter[agent.index]:
                     # TODO: awful late night implementation, improve ASAP
                     self.new_detected_daily += 1
-                # if tested positive then isolate agent
-                if self.consts.should_isolate_positive_detected:
-                    self.update_matrix_manager.apply_full_isolation_on_agent(agent)
 
             agent.set_test_result(test_result)
 
-        # TODO send the detected agents to the selected kind of isolation
-        # TODO: Track isolated agents
         # TODO: Remove healthy agents from isolation?
+        self.isolate_agents()
 
         for new_test in new_tests:
             new_test.agent.set_test_start()
             self.pending_test_results.append(new_test)
 
+    def isolate_agents(self):
+        # those who had the detection test before isolate_after_num_day days exactly
+        can_be_isolated = (self.current_step - self.date_of_last_test) == self.consts.isolate_after_num_day
+        # last test was positive and was the first one
+        tested_positive_once = self.tested_positive_vector & (self.tested_positive_vector_counter == 1)
+        remaining = np.array(self.agents)[can_be_isolated & tested_positive_once]  # get those agents
+        num_of_obedients = round(self.consts.p_will_obey_isolation * len(remaining))  # get number of agents to sample
+        will_obey_isolation = np.random.choice(remaining, num_of_obedients, replace=False)  # sample those who will obey
+
+        for agent in will_obey_isolation:
+            self.agents_in_isolation[agent.index] = True  # keep track about who is in isolation
+            self.update_matrix_manager.change_agent_relations_by_factor(agent,
```

Comment:
What will happen if the matrix will be reset? (e.g., `reset_policies_by_connection_type` will be called).
In that case, I think that this factor will be overridden.

We need to keep track of the isolated agents in the `UpdateMatrix` and prevent the reset method to effectively cancel their isolation

Reply:
fixed